### PR TITLE
GH-40563: [Go] Unable to JSON marshal float64 arrays which contain a NaN value

### PR DESCRIPTION
**Pull request by verma-kartik**
### Rationale for this change

From this [issue](https://github.com/apache/arrow/issues/40563#issuecomment-2043464834).

### What changes are included in this PR?

MarshalJSON() is changed for Float32, Float64 to handle NaN, +Inf and -Inf values. I was only able to handle NaN for Float16. 

UnmarshalJSON() was already handling NaN, +Inf and -Inf in the right way. 

### Are these changes tested?

These changes are tested. Tests are included in those files where I find it was logically appropriate to do so.

### Are there any user-facing changes?

They will be able to handle corner values for Float32, Float64 now.

* GitHub Issue: #40563

### CODE REVIEW

File: go/arrow/array/float16.go [hunk 1]
```diff
@@ -87,10 +87,16 @@ func (a *Float16) GetOneForMarshal(i int) interface{} {
...
 			vals[i] = nil
+			continue
+		}
+
+		switch {
+		case v.IsNaN():
+			vals[i] = "NaN"
+		default:
+			vals[i] = v.Float32()
 		}
```
Comment (zeroshade) [inline go/arrow/array/float16.go:100]: no check for `Inf`?

File: go/arrow/array/numeric.gen.go [hunk 2]
```diff
@@ -575,10 +588,21 @@ func (a *Float32) GetOneForMarshal(i int) interface{} {
...
+		switch {
+		case strconv.FormatFloat(float64(f), 'g', -1, 32) == "NaN":
+            vals[i] = "NaN"
+		case strconv.FormatFloat(float64(f), 'g', -1, 32) == "+Inf":
+            vals[i] = "+Inf"
+		case strconv.FormatFloat(float64(f), 'g', -1, 32) == "-Inf":
+            vals[i] = "-Inf"
+		default:
+			vals[i] = f
 		}
```
Comment (zeroshade) [inline go/arrow/array/numeric.gen.go:606]: Please fix the indentation.

Also, let's only call `strconv.FormatFloat` once: 

```go
switch v := strconv.FormatFloat(float64(f), 'g', -1, 32) {
case "NaN", "+Inf", "-Inf":
  vals[i] = v
default:
  vals[i] = f
}
```

**Comment by zeroshade** [comment]
@verma-kartik for the time being, since we are still using the `.tmpl` files and code gen, please update the corresponding template file for `numeric.gen.go.tmpl` rather than only updating the `.go` file.

File: go/arrow/array/float16.go (hunk 1 above)
Comment (verma-kartik) [reply go/arrow/array/float16.go:100]: Yeah. I had a doubt in this. If I use float16.Inf(), I assume that is "+Inf". What for "-Inf" then?

**Comment by verma-kartik** [comment]
Oh, this looks wrong. Seems like, I pulled upstream on my branch instead of main.

File: go/arrow/array/float16.go (hunk 1 above)
Comment (zeroshade) [reply go/arrow/array/float16.go]: We should add an `IsInf` method to `float16.Num` which can do `return n.bits & 0x7c00`. After that you can simply check the sign bit via `Signbit()` to determine if it's `-Inf` or `+Inf`.

File: go/arrow/array/float16.go [hunk 3]
```diff
@@ -87,10 +87,18 @@ func (a *Float16) GetOneForMarshal(i int) interface{} {
...
 			vals[i] = nil
+			continue
+		}
+
+		switch {
+		case v.IsNaN():
+			vals[i] = "NaN"
+		case v == float16.Inf():
+			vals[i] = "Inf"
+		default:
```
Comment (zeroshade) [inline go/arrow/array/float16.go:100]: Once both positive and negative infinity are handled for float16, I'm fine to merge this

**Review by zeroshade — CHANGES_REQUESTED** [review]

File: go/arrow/array/float16.go (hunk 3 above)
Comment (verma-kartik) [reply go/arrow/array/float16.go:100]: I'll do it soon. Been busy with work.

**Review by zeroshade — APPROVED** [review]
LGTM thanks for this!

Will merge after CI passes
